clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/AbstractPubSubBasedMessageSender.java

```diff
@@ -0,0 +1,306 @@
+        final ByteString data = ByteString.copyFrom(payload.getBytes());
+        final PubsubMessage pubsubMessage = PubsubMessage.newBuilder().putAllAttributes(pubSubAttributes)
+                .setOrderingKey(deviceId).setData(data).build();
+
+        log.trace("sending message to Pub/Sub [topic: {}, registry: {}, deviceId: {}]", topic, tenantId, deviceId);
+        logPubSubMessage(currentSpan, pubsubMessage, topic, tenantId);
+
+        return getOrCreatePublisher(topic, tenantId).publish(pubsubMessage)
+                .onSuccess(recordMessage -> {
+                    logPubSubMessageId(currentSpan, topic, recordMessage);
+                })
+                .onFailure(t -> {
+                    logError(currentSpan, topic, tenantId, deviceId, t);
+                    throw new ServerErrorException(tenantId, getErrorCode(), t);
+                })
+                .mapEmpty();
+
+    }
+
+    /**
+     * Creates a new <em>OpenTracing</em> span to trace publishing messages to Pub/Sub.
+     *
+     * @param operationName The operation name to set for the span.
+     * @param topic The topic to which the message is sent.
+     * @param tenantId The tenant identifier related to the operation.
+     * @param deviceId The device identifier related to the operation.
+     * @param referenceType The type of reference towards the given span context.
+     * @param context The span context to set as parent and to derive the sampling priority from (may be null).
+     * @return The new span.
+     * @throws NullPointerException if tracer or topic is {@code null}.
+     */
+    protected Span startSpan(final String operationName, final String topic, final String tenantId,
+            final String deviceId, final String referenceType, final SpanContext context) {
+        Objects.requireNonNull(operationName);
+        Objects.requireNonNull(topic);
+        Objects.requireNonNull(referenceType);
+        return newSpan(tracer, operationName, topic, referenceType, context)
+                .setTag(TracingHelper.TAG_TENANT_ID.getKey(), tenantId)
+                .setTag(TracingHelper.TAG_DEVICE_ID.getKey(), deviceId);
+    }
+
+    private PubSubPublisherClient getOrCreatePublisher(final String topic, final String tenantId) {
+        return publisherFactory.getOrCreatePublisher(topic, projectId, tenantId);
+    }
+
+    private void logPubSubMessageId(final Span span, final String topic, final String messageId) {
+        log.trace("message published to PubSub [topic: {}, id: {}]", topic, messageId);
+        span.log("message published to PubSub");
+
+        Tags.HTTP_STATUS.set(span, HttpURLConnection.HTTP_ACCEPTED);
+    }
+
+    private void logPubSubMessage(final Span span, final PubsubMessage message, final String topic,
+            final String tenantId) {
+        final String attributesAsString = message.getAttributesMap()
+                .keySet()
```

Comment:
how about using `Map.entrySet()` in order to prevent looking up each value individually?

Reply:
Good point, changed it to use `Map.entrySet()`